diff -up ./cpupower/man/cpupower-set.1.orig ./cpupower/man/cpupower-set.1
--- ./cpupower/man/cpupower-set.1.orig	2013-02-22 10:28:03.518913123 -0500
+++ ./cpupower/man/cpupower-set.1	2013-02-22 10:29:07.755711873 -0500
@@ -3,7 +3,7 @@
 cpupower\-set \- Set processor power related kernel or hardware configurations
 .SH SYNOPSIS
 .ft B
-.B cpupower set [ \-b VAL ] [ \-s VAL ] [ \-m VAL ]
+.B cpupower set [ \-b VAL ] [ \-s VAL ]
 
 
 .SH DESCRIPTION
@@ -56,12 +56,9 @@ Use \fBcpupower -c all info -b\fP to ver
 This options needs the msr kernel driver (CONFIG_X86_MSR) loaded.
 .RE
 .PP
-\-\-sched\-mc,  \-m [ VAL ]
 .RE
 \-\-sched\-smt, \-s [ VAL ]
 .RS 4
-\-\-sched\-mc utilizes cores in one processor package/socket first before
-processes are scheduled to other processor packages/sockets.
 
 \-\-sched\-smt utilizes thread siblings of one processor core first before
 processes are scheduled to other cores.
diff -up ./cpupower/utils/cpupower-set.c.orig ./cpupower/utils/cpupower-set.c
--- ./cpupower/utils/cpupower-set.c.orig	2013-02-22 09:39:52.401942275 -0500
+++ ./cpupower/utils/cpupower-set.c	2013-02-22 10:27:43.071340512 -0500
@@ -19,7 +19,6 @@
 
 static struct option set_opts[] = {
 	{ .name = "perf-bias",	.has_arg = optional_argument,	.flag = NULL,	.val = 'b'},
-	{ .name = "sched-mc",	.has_arg = optional_argument,	.flag = NULL,	.val = 'm'},
 	{ .name = "sched-smt",	.has_arg = optional_argument,	.flag = NULL,	.val = 's'},
 	{ },
 };
@@ -52,7 +51,7 @@ int cmd_set(int argc, char **argv)
 
 	params.params = 0;
 	/* parameter parsing */
-	while ((ret = getopt_long(argc, argv, "m:s:b:",
+	while ((ret = getopt_long(argc, argv, "s:b:",
 						set_opts, NULL)) != -1) {
 		switch (ret) {
 		case 'b':
@@ -66,17 +65,6 @@ int cmd_set(int argc, char **argv)
 			}
 			params.perf_bias = 1;
 			break;
-		case 'm':
-			if (params.sched_mc)
-				print_wrong_arg_exit();
-			sched_mc = atoi(optarg);
-			if (sched_mc < 0 || sched_mc > 2) {
-				printf(_("--sched-mc param out "
-					 "of range [0-%d]\n"), 2);
-				print_wrong_arg_exit();
-			}
-			params.sched_mc = 1;
-			break;
 		case 's':
 			if (params.sched_smt)
 				print_wrong_arg_exit();
